internal/login/login.go

```diff
@@ -1,18 +1,225 @@
+
+		if err != nil {
+			return accessTokenResponse, fmt.Errorf("cannot read access token response body: %w", err)
+		}
+
+		if err := json.Unmarshal(body, &accessTokenResponse); err != nil {
+			return accessTokenResponse, fmt.Errorf("cannot unmarshal access token response: %w", err)
+		}
+
+		return accessTokenResponse, nil
+	}
+
+	return accessTokenResponse, fmt.Errorf("HTTP %s: cannot retrieve access token", resp.Status)
+}
+
+func Run(stdin *os.File, params RunParams) error {
+	if params.Token != "" {
+		err := utils.SaveAccessToken(params.Token, params.Fsys)
+		if err != nil {
+			return fmt.Errorf("cannot save provided token: %w", err)
+		}
+		fmt.Println(loggedInMsg)
+		return nil
+	}
+
+	if params.OpenBrowser {
+		fmt.Print("Hello from ", utils.Aqua("Supabase"), "! Press ", utils.Aqua("Enter"), " to open browser and login automatically.\n")
+		fmt.Scanln()
+	}
+
+	sessionId := uuid.New().String()
+	tokenName := params.Name
+	curve := ecdh.P256()
+	privateKey, err := curve.GenerateKey(rand.Reader)
+	if err != nil {
+		return fmt.Errorf("cannot generate encryption key: %w", err)
+	}
+	publicKey := privateKey.PublicKey()
+	encodedPublicKey := hex.EncodeToString(publicKey.Bytes())
+
+	createLoginSessionPath := "/cli/login"
+	createLoginSessionQuery := "?session_id=" + sessionId + "&token_name=" + tokenName + "&public_key=" + encodedPublicKey
+	createLoginSessionUrl := utils.GetSupabaseDashboardURL() + createLoginSessionPath + createLoginSessionQuery
+
+	if params.OpenBrowser {
+		fmt.Printf("Here is your login link in case browser did not open: %s\n\n", utils.Bold(createLoginSessionUrl))
+
+		openCmd := exec.Command("open", createLoginSessionUrl)
+		if err := openCmd.Run(); err != nil {
+			return fmt.Errorf("cannot open default browser: %w", err)
+		}
+	} else {
+		fmt.Printf("Here is your login link: %s\n\n", utils.Bold(createLoginSessionUrl))
+	}
+
+	pollingProcess := tea.NewProgram(pollingState{
+		message: "Your token is now being generated and securely encrypted. Waiting for it to arrive...",
+		spinner: utils.CreateSpinner(),
+	})
+
+	tokenCh := make(chan AccessTokenResponse, 1)
+	errCh := make(chan error, 1)
+
+	go func() {
+		sessionPollingUrl := utils.GetSupabaseAPIHost() + "/platform/cli/login/" + sessionId
+		accessTokenResponse, err := pollForAccessToken(sessionPollingUrl)
+		if err != nil {
+			errCh <- err
+		} else {
+			tokenCh <- accessTokenResponse
+		}
+		pollingProcess.Quit()
+	}()
+
+	if _, err := pollingProcess.Run(); err != nil {
+		return fmt.Errorf("cannot render polling status: %w", err)
+	}
+
+	select {
+	case err := <-errCh:
+		return err
+	case accessTokenResponse := <-tokenCh:
+		decryptedAccessToken, err := decryptAccessToken(accessTokenResponse, curve, privateKey)
+		if err != nil {
+			return err
+		}
+
+		fmt.Printf("Token created successfully: %s\n\n", utils.Bold(tokenName))
+		fmt.Println(loggedInMsg)
+
+		return utils.SaveAccessToken(decryptedAccessToken, params.Fsys)
+	}
```

Comment:
```suggestion
	return utils.RunProgram(ctx, func(p utils.Program, ctx context.Context) {
		p.Send(utils.StatusMessage("Your token is now being generated and securely encrypted. Waiting for it to arrive..."))

		sessionPollingUrl := utils.GetSupabaseAPIHost() + "/platform/cli/login/" + sessionId
		accessTokenResponse, err := pollForAccessToken(sessionPollingUrl)
		if err != nil {
			return err
		}
		decryptedAccessToken, err := decryptAccessToken(accessTokenResponse, curve, privateKey)
		if err != nil {
			return err
		}

		fmt.Printf("Token created successfully: %s\n\n", utils.Bold(tokenName))
		fmt.Println(loggedInMsg)

		return utils.SaveAccessToken(decryptedAccessToken, params.Fsys)
	})
```

I think something like this would work? Not much changes given our current refactoring.

Reply:
Perfection :chefs_kiss:

Reply:
Had to wait for program to end first, otherwise prints overlap, as bubbletea controls terminal state inside the program.